File: Kepler's Laws Simulations/Kepler's Laws.py
import math

G = 6.6743e-11
M = 1.9891e30
Au = 1.496e+11
dt = 30

time = 31536000  

class Planet:
    def __init__(self, mass, vel, pos):
        self.mass = mass
        self.vel = vel  # [vx, vy]
        self.pos = pos  # [x, y]

def force(r, mass):
    F = (G * M * mass) * (1 / r**2)
    return -F

def radius(x, y):
    return math.sqrt(x**2 + y**2)

def area(r, theta):
    return 0.5 * (r**2) * theta

# Initialize planets
earth = Planet(5.97219e24, [0, 30e3], [Au, 0])
mars = Planet(6.39e23, [0, 24e3], [1.5 * Au, 0])

earth_data = []
mars_data = []

for t in range(0, time, dt):
    theta0 = math.atan2(earth.pos[1], earth.pos[0])
    
    r = radius(earth.pos[0], earth.pos[1])
    rm = radius(mars.pos[0], mars.pos[1])
    
    a = force(r, earth.mass) / earth.mass
    am = force(r, mars.mass) / mars.mass
    
    ax = a * (earth.pos[0] / r)
    ay = a * (earth.pos[1] / r)
    axm = am * (mars.pos[0] / r)
    aym = am * (mars.pos[1] / r)
    
    earth.vel[0] += ax * dt
    earth.vel[1] += ay * dt
    
    mars.vel[0] += axm * dt
    mars.vel[1] += aym * dt
    
    earth.pos[0] += (earth.vel[0] * dt) + 0.5 * (ax * dt**2)
    earth.pos[1] += (earth.vel[1] * dt) + 0.5 * (ay * dt**2)
    
    mars.pos[0] += (mars.vel[0] * dt) + 0.5 * (axm * dt**2)
    mars.pos[1] += (mars.vel[1] * dt) + 0.5 * (aym * dt**2)
    
    theta = math.atan2(earth.pos[1], earth.pos[0])
    theta = theta - theta0
    area_value = area(r, theta)
    
    earth_data.append(f"{t} {earth.pos[0]} {earth.pos[1]} {area_value}\n")
    mars_data.append(f"{t} {mars.pos[0]} {mars.pos[1]} {area_value}\n")

with open("earth.txt", "w") as earth_file:
    earth_file.writelines(earth_data)

with open("mars.txt", "w") as mars_file:
    mars_file.writelines(mars_data)

File: Kepler's Laws Simulations/Kepler's Laws.cpp
/*
 Author: Haroon Sakhi
 Description: This code was developed by Haroon Sakhi as part
 of the project Kepler's Laws Simulations.
 Any use or modification of this code should include proper
 attribution to the original author.
*/

#include <iostream>
#include <fstream>
#include <math.h>

using namespace std;

double G = 6.6743e-11;
double M = 1.989e30;
double Au = 1.496e+11;
double dt = 30;

class Planet
{
public:
    double mass;
    double vel[2];
    double pos[2];
    double acc[2];
};

double Force(double r, double mass)
{
    double F = (G * M * mass) * (1 / pow(r, 2));
    return -F;
}

double Radius(double x, double y)
{
    double R = sqrt((x * x) + (y * y));
    return R;
}

double Area(double r, double theta)
{
    double a = 0.500 * (r * r) * theta;
    return a;
}

int main()
{
    ofstream earth("earth.txt");
    ofstream mars("mars.txt");

    Planet planets[2];

    planets[0].mass = 5.97219e24;
    planets[0].vel[0] = 0;
    planets[0].vel[1] = 30e3;
    planets[0].pos[0] = Au;
    planets[0].pos[1] = 0;

    planets[1].mass = 6.39e23;
    planets[1].vel[0] = 0;
    planets[1].vel[1] = 24e3;
    planets[1].pos[0] = 1.500 * Au;
    planets[1].pos[1] = 0;

    Planet Mars = planets[1];
    Planet Earth = planets[0];

    double time = 31536000;
    double a, theta0, theta, r, ax, ay;
    double am, rm, axm, aym;
    double area = 0;
    theta0 = 0;
    theta = 0;
    area = 0;

    for (int t = 0; t < time; t = t + dt)
    {
        theta0 = atan((Earth.pos[1]) / (Earth.pos[0]));

        earth << t << " " << Earth.pos[0] << " " << Earth.pos[1] << " " << area << endl;
        mars << t << " " << Mars.pos[0] << " " << Mars.pos[1] << " " << area << endl;

        r = Radius(Earth.pos[0], Earth.pos[1]);
        rm = Radius(Mars.pos[0], Mars.pos[1]);
        a = Force(r, Earth.mass) / Earth.mass;
        am = Force(rm, Mars.mass) / Mars.mass;

        ax = a * (Earth.pos[0] / r);
        ay = a * (Earth.pos[1] / r);
        axm = am * (Mars.pos[0] / r);
        aym = am * (Mars.pos[1] / r);

        Earth.vel[0] += ax * dt;
        Earth.vel[1] += ay * dt;
        Mars.vel[0] += axm * dt;
        Mars.vel[1] += aym * dt;

        Earth.pos[0] += (Earth.vel[0] * dt) + 0.500 * (ax * pow(dt, 2));
        Earth.pos[1] += (Earth.vel[1] * dt) + 0.500 * (ay * pow(dt, 2));

        Mars.pos[0] += (Mars.vel[0] * dt) + 0.500 * (axm * pow(dt, 2));
        Mars.pos[1] += (Mars.vel[1] * dt) + 0.500 * (aym * pow(dt, 2));

        theta = atan((Earth.pos[1]) / (Earth.pos[0]));
        theta = theta - theta0;
        area = Area(r, theta);
    }

    return 0;
}
